### PR TITLE
Sb rb connect to firestore

### PR DESCRIPTION
**Pull request by shirenekboyd**
## Description

We installed `firebase` as a project dependency in our package.json file.  We added two new components, one for adding/sending data to the Cloud Firestore database with a `<button/>`, and another for getting/listening to data from Firestore and displaying it.  We are rendering these components (AddData.js, ShowData.js) in the App component.

## Related Issue

closes 1

## Acceptance Criteria

- [ ✓] `firebase` has been added as a project dependency
- [✓ ] Making a change in the Firestore database will update the data in the app
- [✓ ] Making a change in the app will update the data in Firestore

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|  ✓  | :link: Update dependencies |
|    | :scroll: Docs              |

### After

Screen Shot 2022-04-07 at 3 01 21 PM

## Testing Steps / QA Criteria

- Pull down the branch using `git pull origin sb-rb-connect-to-firestore`.
- `git checkout sb-rb-connect-to-firestore` to switch branches.
- `npm start` to open a local copy of the app.
- Click the button to add sample data to the Cloud Firestore database.

### CODE REVIEW

File: src/components/ShowData.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    return () => unsubscribe();
+  }, []);
+
+  //   console.log('this is docs--', docs);
+
+  return (
+    <div>
+      {docs.map((doc) => {
+        return (
+          <ul>
```
Comment (timothy-taylor) [inline src/components/ShowData.js]: we'll want to move `<ul>` outside of the map function here

File: src/components/ShowData.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+            <li>{doc.id}</li>
```
Comment (timothy-taylor) [inline src/components/ShowData.js]: as per [React docs](https://reactjs.org/docs/lists-and-keys.html), we'll want to add a key to list elements `<li key={doc.id}>`

**Review by timothy-taylor — CHANGES_REQUESTED** [review]
Looks good! I left code comments inline for a couple minor syntax things around the map function in ShowData.  

One thought I had, and this might be excessive, but: I wonder whether we want to think about separating our backend logic from our frontend UI at this stage. The bulk of AddData and ShowData is really logic that could be made into custom hooks, like `useSnapshot` and `useAddDoc`, which could then be imported into our UI components as necessary.

for example, useSnapshot:
```
export const useSnapshot = () => { 
  const [docs, setDocs] = useState([]);

  useEffect(() => { 
    const unsubscribe = onSnapshot(collection(db, 'users'), (snapshot) => { 
      const snapshotDocs = [];
      snapshot.forEach((doc) => snapshotDocs.push(doc));
      setDocs(snapshotDocs);
    });
    return () => unsubscribe();
  }, []);

  return docs;
}
```

Since we (probably) won't need to use these logic hooks across multiple components this might not be necessary, so don't feel like you have to do this. I was just thinking about separation of concerns, and wanted to throw the idea out there in case it was interesting to anyone else.

File: src/components/ShowData.js (hunk 2 above)
Comment (rjwbeard) [reply src/components/ShowData.js]: thanks for noticing that we missed this! I added a key attribute to our list items

File: src/components/ShowData.js (hunk 1 above)
Comment (rjwbeard) [reply src/components/ShowData.js]: fixed it! another great catch

**Comment by rjwbeard** [comment]
@timothy-taylor thank you for your thoughtful feedback! I've resolved the syntax errors locally on my machine and I'm wondering how to proceed in regards to file structure. I like your idea of separating the back end logic from our UI components and since we've established that our components are reading from and writing to Firestore, I can do that refactor now and remove the JSX from our components so they're ready for future use. I can rename the folder for these refactored files (useSnapshot and useAddDoc) to reflect that they're custom hooks.

**Comment by rjwbeard** [comment]
I'm realizing now that I'm removing the JSX that I'm also deleting those syntax changes you suggested and that I just made. Nevertheless I thank you for catching them, it's still good practice to go back and make those edits!

File: src/hooks/useAddDoc.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { useState } from 'react';
+import { db } from '../lib/firebase';
+import { collection, addDoc } from 'firebase/firestore';
+
+export default function useAddDoc() {
```
Comment (timothy-taylor) [inline src/hooks/useAddDoc.js]: We'll need to include the `itemName` variable used in line 12 as an argument to the function.

**Review by timothy-taylor — DISMISSED** [review]
Looks good. Good call on the `hooks` directory. There is just one little missing argument in useAddDoc. 

It would be awesome to be able to test these hooks at some point. That is not trivial at all though, but could be a fun thing to read up on in our spare time.

**Pull request closed without merging**